dimensionNarrowing.py

```diff
@@ -53,10 +55,64 @@
                    'Primary energy consumption per capita (kWh/person)', 'Value_co2_emissions_kt_by_country',
                    'gdp_growth', 'gdp_per_capita', 'Density\\n(P/Km2)', 'Land Area(Km2)']
 
-# Min-Max normalization for numeric columns
+# Task 3: Min-Max normalization for numeric columns
 min_vals = filtered_data[numeric_columns].min()
 max_vals = filtered_data[numeric_columns].max()
 normalized_data_min_max[numeric_columns] = (
     filtered_data[numeric_columns] - min_vals) / (max_vals - min_vals)
 
-normalized_data_min_max.to_csv('normalized.csv', index=False)
\ No newline at end of file
+normalized_data_min_max.to_csv('normalized.csv', index=False)
+
+# Task 4: dimension narrowing
+
+# UMAP method
+
+# filtered_data_grouped_by_country = filtered_data.groupby("Entity").mean()
+filtered_data_grouped_by_country = filtered_data.groupby(
+    "Entity", as_index=False).mean()
+filtered_data_grouped_by_country.to_csv(
+    'filtered_data_grouped_by_country.csv', index=False)
+
+
+umap_data = filtered_data_grouped_by_country.drop(
+    columns=["Entity", "Access to electricity (% of population)"])
+
+reduced_data_umap = umap.UMAP(n_components=2, random_state=42).fit_transform(
+    umap_data)
+
+umap_df = pd.DataFrame(reduced_data_umap, columns=["x", "y"])
+umap_df["Entity"] = filtered_data_grouped_by_country["Entity"]
+umap_df["Access to electricity (% of population)"] = filtered_data_grouped_by_country["Access to electricity (% of population)"]
+
+# Use Plotly Express for interactive plotting
+fig_electricity_percentage = px.scatter(umap_df, x="x", y="y",
+                                        color="Access to electricity (% of population)", hover_name="Entity")
+# Adjust marker size for better visibility
+fig_electricity_percentage.update_traces(marker=dict(size=5))
+fig_electricity_percentage.update_layout(title="UMAP Projection")
+fig_electricity_percentage.show()
+
+normalized_data_grouped_by_country = normalized_data_min_max.groupby(
+    "Entity", as_index=False).mean()
+
+# UMAP for normalized data
```

Comment:
Is there a UMAP for non-normalized data if so please add a comment like this. If not this could be rephrased